toolchain/check/member_access.cpp

```diff
@@ -155,23 +155,33 @@ static auto LookupInterfaceWitness(Context& context,
 
 // Performs impl lookup for a member name expression. This finds the relevant
 // impl witness and extracts the corresponding impl member.
-static auto PerformImplLookup(Context& context, Parse::NodeId node_id,
-                              SemIR::ConstantId type_const_id,
-                              SemIR::AssociatedEntityType assoc_type,
-                              SemIR::InstId member_id) -> SemIR::InstId {
+static auto PerformImplLookup(
+    Context& context, SemIR::LocId loc_id, SemIR::ConstantId type_const_id,
+    SemIR::AssociatedEntityType assoc_type, SemIR::InstId member_id,
+    std::optional<Context::Diagnoser> missing_impl_diagnoser) -> SemIR::InstId {
   auto interface_type =
       context.types().GetAs<SemIR::InterfaceType>(assoc_type.interface_type_id);
   auto& interface = context.interfaces().Get(interface_type.interface_id);
   auto witness_id = LookupInterfaceWitness(context, type_const_id,
                                            assoc_type.interface_type_id);
   if (!witness_id.is_valid()) {
-    CARBON_DIAGNOSTIC(MissingImplInMemberAccess, Error,
-                      "Cannot access member of interface {0} in type {1} "
-                      "that does not implement that interface.",
-                      SemIR::NameId, std::string);
-    context.emitter().Emit(node_id, MissingImplInMemberAccess,
-                           interface.name_id,
-                           context.sem_ir().StringifyType(type_const_id));
+    if (missing_impl_diagnoser) {
+      CARBON_DIAGNOSTIC(MissingImplInMemberAccessNote, Note,
+                        "Type `{1}` does not implement interface `{0}`.",
+                        SemIR::NameId, std::string);
+      (*missing_impl_diagnoser)()
+          .Note(loc_id, MissingImplInMemberAccessNote, interface.name_id,
+                context.sem_ir().StringifyType(type_const_id))
+          .Emit();
+    } else {
+      CARBON_DIAGNOSTIC(MissingImplInMemberAccess, Error,
+                        "Cannot access member of interface `{0}` in type `{1}` "
+                        "that does not implement that interface.",
+                        SemIR::NameId, std::string);
+      context.emitter().Emit(loc_id, MissingImplInMemberAccess,
+                             interface.name_id,
+                             context.sem_ir().StringifyType(type_const_id));
+    }
```

Comment:
Can these use TypeId directly?

```suggestion
    if (missing_impl_diagnoser) {
      CARBON_DIAGNOSTIC(MissingImplInMemberAccessNote, Note,
                        "Type `{1}` does not implement interface `{0}`.",
                        SemIR::NameId, SemIR::TypeId);
      (*missing_impl_diagnoser)()
          .Note(loc_id, MissingImplInMemberAccessNote, interface.name_id,
               type_const_id)
          .Emit();
    } else {
      CARBON_DIAGNOSTIC(MissingImplInMemberAccess, Error,
                        "Cannot access member of interface `{0}` in type `{1}` "
                        "that does not implement that interface.",
                        SemIR::NameId, SemIR::TypeId);
      context.emitter().Emit(loc_id, MissingImplInMemberAccess,
                             interface.name_id,
                             type_const_id);
    }
```

Reply:
Done. (`type_const_id` is a `ConstantId` not a `TypeId`, so the suggestion doesn't quite work.)